libraries/rush-lib/src/logic/operations/Operation.ts

```diff
@@ -148,4 +154,83 @@ export class Operation {
     (this.dependencies as Set<Operation>).delete(dependency);
     (dependency.consumers as Set<Operation>).delete(this);
   }
+
+  public get stateHash(): string {
+    if (!this._stateHash) {
+      throw new Error(
+        'Operation state hash is not calculated yet, you must call `calculateStateHash` first.'
+      );
+    }
+    return this._stateHash;
+  }
+
+  public calculateStateHash(options: {
+    inputsSnapshot: IInputsSnapshot;
+    buildCacheConfiguration: BuildCacheConfiguration;
+  }): string {
+    if (this._stateHash) {
+      return this._stateHash;
+    }
+    const { inputsSnapshot, buildCacheConfiguration } = options;
+    const { cacheHashSalt } = buildCacheConfiguration;
+    // This redefinition is necessary due to limitations in TypeScript's control flow analysis, due to the nested closure.
+    const definitelyDefinedInputsSnapshot: IInputsSnapshot = inputsSnapshot;
+
+    // Build cache hashes are computed up front to ensure stability and to catch configuration errors early.
+    function getOrCreateOperationHash(operation: Operation): string {
+      if (operation._stateHash) {
+        return operation._stateHash;
+      }
+      // Examples of data in the config hash:
+      // - CLI parameters (ShellOperationRunner)
+      const configHash: string | undefined = operation.runner?.getConfigHash();
```

Comment:
In `OperationExecutionRecord`, `runner` cannot be `undefined`, so this gets simpler.